README.md
Simple program that parses the selected excel file.

estimate.sql is just an example of how the table of contets should look.

If you want to test the program, run 'testing_excel_parser.php'.

ReleaseNotes.txt

----------Release Notes------------

v. 1.0.0
Changes: 
	Removed db use
	Unit tests are now written to file
	Added release notes
	Removed die use
	Minor code changes

v. 1.0.1
Changes:
	function now takes 2 parameters: 1st - the file that will be parsed.
					2nd - a number(1 - if running parser noramly, prints all the info on screen
							2 - if running unit testing, doesnt print anything on screen)


v. 2.0.0
Changes:
	jobs_array fields have been renamed
	added new estimate_sum_info array
	excel_parser now uses excel_parser_function
	function now reuturns an array containing both jobs_array and estimate_sum_array
	
	

test/test_parse.php
<?php
function main_main($file)
{
    // Enter your mysql username
    $user = 'root';
    // Enter your mysql password
    $password = '';
    // Enter the database you are going to use
    $database = 'test';
    //Default host
    $host = 'localhost';

    // Connect to mysql
    $con = mysqli_connect($host, $user, $password, $database);
    $con->set_charset("utf8");

    // Check if there was an error connecting
    if ($con->connect_error)
    {
        die('Connect Error (' . $con->connect_errno . ') '
                . $con->connect_error);
    }
    
    //Check if table exists, if yes - drop it.
    if ($con->query("SHOW TABLES LIKE 'Estimate'")){$con->query("DROP TABLE Estimate");}

    //$file = $_FILES['excel']['name'];
    if(isset($file))
    {   
        //Check if user inputed the correct file type
        $wanted = array('xlsx');
        $check_extension = pathinfo($file, PATHINFO_EXTENSION);
        if (!in_array($check_extension, $wanted))
        {
            //echo "You have inputed the wrong file type <br>";
            //echo "Wanted file type: xlsx(excel)";
            $file = NULL;
        }else
        {
            
            //Create table
            $sql = "CREATE TABLE Estimate(
                Id INT(6) UNSIGNED AUTO_INCREMENT PRIMARY KEY,
                Work_name VARCHAR(150) COLLATE 'utf8_general_ci',
                Material_type VARCHAR(50) COLLATE 'utf8_general_ci',
                Mass VARCHAR(20) COLLATE 'utf8_general_ci',
                Mass_m VARCHAR(20) COLLATE 'utf8_general_ci',
                Work VARCHAR(20) COLLATE 'utf8_general_ci',
                Work_m VARCHAR(20) COLLATE 'utf8_general_ci'
            )";
            
            $result = mysqli_query($con, $sql) or die ("Bad create: $con->error");
            
            //Incude excel_parser class
            require_once __DIR__.'../../excel_parser_class.php';
            
            $excel = SimpleXLSX::parse($file);
            
            $num_cols = 15;
            $num_rows = 200;

            $invalid_val = -1;
            $work_col_numb = $invalid_val;
            $mass_col_numb = $invalid_val;
            $measurement_col_numb = $invalid_val;

            $rows = $excel->rows();
            $crr_row_idx = 0;
            $jobs_found = 0;

            //echo 'Starting! <br>';
            //Start parsing
            foreach ($rows as $c)
            {   
                // Exit causes
                if ($crr_row_idx >= $num_rows)
                {
                    //echo "<br>Exited by reaching maxmimum number of rows! <br>";
                    break;
                }
                if ($c[1] == 'Iš viso')
                {
                    //echo "<br>Exited by reaching the conclusion! <br>";
                    break;
                }

                $name = '';
                $material = '';
                $mass = '';
                $mass_m = '';
                $work = '';
                $work_m = ''; 

                //Find header
                if ($c[0] == 'Eil. Nr.')
                {
                    //echo "Found header! <br>";
                    $y = 0;
                    for ($y; $y <= $num_cols; $y++)
                    {  
                        // Find measurement column in the header
                        if (isset($c[$y]) && $c[$y] == 'Mato vnt.')
                        {
                            $measurement_col_numb = $y;
                            continue;
                        }else {}
                        // Find work column in the header
                        if (isset($c[$y]) && $c[$y] == 'Darbas')
                        {
                            $work_col_numb = $y;
                            continue;
                        }else{}
                        // Find mass column in the header
                        if (isset($c[$y]) && $c[$y] == 'Medžiagos')
                        {
                            $mass_col_numb = $y;
                            continue;
                        }else{}
                    }
                }
                
                //Check if summaries first cell has a digit
                if (ctype_digit($c[0]))
                {
                    //Chekc if header was found
                    if ($work_col_numb == $invalid_val)
                    {
                        //echo 'Header not found';
                        break;
                    }
                    //echo "<br>Found job! <br>";
                    $jobs_found++;
                    //Assign name variable
                    $name = $c[1];
                    //print_r($name);
                    //echo '<br>';
                    $next_row_idx = $crr_row_idx+1;
                    $next_row = $rows[$next_row_idx];

                    //Find and assign work variables
                    if ($next_row[0] == '' && $next_row[1] == 'Darbas')
                    {
                        $work = $next_row[$work_col_numb];
                        $work_m = $next_row[$measurement_col_numb];

                    }//else{echo "No work data <br>";}
                    
                    $next_row_idx++;
                    $next_row = $rows[$next_row_idx];

                    //Find and assign material variables
                    if ($next_row[0] == '' && $next_row[$mass_col_numb] != '')
                    {
                        $material = $next_row[1];
                        $mass = $next_row[$mass_col_numb];
                        $mass_m = $next_row[$measurement_col_numb];
                        
                    }//else{echo "No material data <br>";}

                    //Insert into mysqli table
                    $estimate = 'INSERT INTO estimate (Work_name, Material_type, Mass, Mass_m, Work, Work_m) VALUES 
                                                        ("'.$name.'", "'.$material.'", "'.$mass.'", "'.$mass_m.'", "'.$work.'", "'.$work_m.'")';

                    if (mysqli_query($con, $estimate)){}
                }else{}
                
                
                $crr_row_idx++;
            }
            //echo '<br>Finished!<br>';
            //echo 'Found '.$jobs_found.' jobs!';
            //Close connection
            mysqli_close($con);
        }
    }
}
?>


test/testing_excel_parser.php
<?php
require_once __DIR__.'../../excel_parser.php';

$crr_date = date("Y-m-d Hi");
$txt_file = 'TestReport.txt';
$test_report = fopen($txt_file, 'w');

fwrite($test_report, $crr_date."\n");

//TESTING EXCEL PARSER
// Test1 Good output
function TC1()
{
    global $test_report, $txt_file;

    fwrite($test_report, "----TEST1----\nImport valid file, expecting 12 lines\n");
    $wanted = 12;
    $got = 0; 
    $filename = "test1.xlsx";
    $return = parse_excel_file($filename, 0);
    $got = count($return);
    if ($wanted == $got)
    {
        fwrite($test_report, "Success, found number of lines ".$got."\n\n");
    }else
    {
        fwrite($test_report, "Failed, found number of lines ".$got."\n\n");
    }
    
}

// TC2 Missing header
function TC2()
{
    global $test_report, $txt_file;

    fwrite($test_report, "----TEST2----\nImport file has no header, expecting no lines\n");
    $wanted = 0;
    $got = 0; 
    $filename = "test2.xlsx";
    $return = parse_excel_file($filename, 0);
    $got = count($return);
    if ($wanted == $got)
    {
        fwrite($test_report, "Success, found number of lines ".$got."\n\n");
    }else
    {
        fwrite($test_report, "Failed, found number of lines ".$got."\n\n");
    }
}

// TC3 Missing summary
function TC3()
{
    global $test_report, $txt_file;

    fwrite($test_report, "----TEST3----\nImport file is missing 1 summary, expecting 11 lines\n");
    $wanted = 11;
    $got = 0; 
    $filename = "test3.xlsx";
    $return = parse_excel_file($filename, 0);
    $got = count($return);
    if ($wanted == $got)
    {
        fwrite($test_report, "Success, found number of lines ".$got."\n\n");
    }else
    {
        fwrite($test_report, "Failed, found number of lines ".$got."\n\n");
    }
}

// TC4 Missing material
function TC4()
{
    global $test_report, $txt_file;

    fwrite($test_report, "----TEST4----\nImport files 1 job missing material row, expecting 12 lines\n");
    $wanted = 12;
    $got = 0; 
    $filename = "test4.xlsx";
    $return = parse_excel_file($filename, 0);
    $got = count($return);
    if ($wanted == $got)
    {
        fwrite($test_report, "Success, found number of lines ".$got."\n\n");
    }else
    {
        fwrite($test_report, "Failed, found number of lines ".$got."\n\n");
    }
}

// TC5 Missing work
function TC5()
{
    global $test_report, $txt_file;

    fwrite($test_report, "----TEST5----\nImport files 1 job missing work row, expecting 11 lines\n");
    $wanted = 11;
    $got = 0; 
    $filename = "test5.xlsx";
    $return = parse_excel_file($filename, 0);
    $got = count($return);
    if ($wanted == $got)
    {
        fwrite($test_report, "Success, found number of lines ".$got."\n\n");
    }else
    {
        fwrite($test_report, "Failed, found number of lines ".$got."\n\n");
    }
}

// TC6 Missing exit condition
function TC6()
{
    global $test_report, $txt_file;

    fwrite($test_report, "----TEST6----\nImport valid file, missing 1 exit condition, expecting 12 lines\n");
    $wanted = 12;
    $got = 0; 
    $filename = "test6.xlsx";
    $return = parse_excel_file($filename, 0);
    $got = count($return);
    if ($wanted == $got)
    {
        fwrite($test_report, "Success, found number of lines ".$got."\n\n");
    }else
    {
        fwrite($test_report, "Failed, found number of lines ".$got."\n\n");
    }
}

// TC7 Wrong extension
function TC7()
{
    global $test_report, $txt_file;

    fwrite($test_report, "----TEST7----\nImport wrong file extension, expecting nothing\n");
    $wanted = 0;
    $got = 0; 
    $filename = "test7.xlsx";
    $return = parse_excel_file($filename, 0);
    $got = count($return);
    if ($wanted == $got)
    {
        fwrite($test_report, "Success, found number of lines ".$got."\n\n");
    }else
    {
        fwrite($test_report, "Failed, found number of lines ".$got."\n\n");
    }
}

// TC8 Correct extension, meesed up file
function TC8()
{
    global $test_report, $txt_file;

    fwrite($test_report, "----TEST8----\nImport valid file, cant get data from inside, expecting no lines\n");
    $wanted = 0;
    $got = 0; 
    $filename = "test8.xlsx";
    $return = parse_excel_file($filename, 0);
    $got = count($return);
    if ($wanted == $got)
    {
        fwrite($test_report, "Success, found number of lines ".$got."\n\n");
    }else
    {
        fwrite($test_report, "Failed, found number of lines ".$got."\n\n");
    }
}

TC1();
TC2();
TC3();
TC4();
TC5();
TC6();
TC7();
TC8();
fclose($test_report);
//$new_name = "TestReport ".$crr_date;
//$test_report = rename($txt_file, $new_name.'.txt');
?>

excel_parser_function.php
<?php
    //Function prints only if $file_specifier is set to 1
    function printer($f, $s)
    {
        if ($s == 1){echo $f;} 
    }

    function parse_excel_file($file, $file_specifier)
    {
        
        printer('Now parsing '.$file.' file!<br><br>', $file_specifier);
        if(isset($file))
        {   
            //Check if user inputed the correct file type
            $wanted = array('xlsx');
            $check_extension = pathinfo($file, PATHINFO_EXTENSION);
            if (!in_array($check_extension, $wanted))
            {
                printer("You have inputed the wrong file type <br>", $file_specifier);
                printer("Wanted file type: xlsx(excel)",  $file_specifier);
                $file = NULL;
            }else
            {
                //Incude excel_parser class
                require_once __DIR__.'/excel_parser_class.php';
                
                // $special_char = array("ą", "č", "ę", "ė", "į", "š", "ų", "ū", "ž", " ", "(", ")");
                // $correct_word = str_replace($special_char, "", $file);
                // $file = $correct_word;
                // echo $file;

                if ($excel = SimpleXLSX::parse($file))
                {
                    $rows = $excel->rows();
                }else
                {
                    printer('Cant parse!', $file_specifier);
                    return NULL;
                }
            
                $num_cols = 15;
                $num_rows = 200;

                $invalid_val = -1;
                $work_col_numb = $invalid_val;
                $material_amoount_col_numb = $invalid_val;
                $measurement_col_numb = $invalid_val;
                $price_col_numb = $invalid_val;
                $amount = $invalid_val;
                $queue_numb = $invalid_val;
                $mecha1 = $invalid_val;
                $mecha2 = $invalid_val;
                
                $crr_row_idx = 0;
                $jobs_found = 0;
                $jobs_array = [];
                $estimate_sum_array = [];

                $task_group_name = '';
                $work_total_sum = '';
                $material_total_sum = '';
                $tools_total_sum = '';
                $tools2_total_sum = '';

                $expences1 = '';
                $expences2 = '';
                $expences3 = '';
                $profit = '';
                $total_income = '';
                $sum = '';

                printer('Starting! <br>', $file_specifier);
                //Start parsing
                foreach ($rows as $c)
                {   
                    // Exit causes
                    if ($crr_row_idx >= $num_rows)
                    {
                        printer("<br>Exited by reaching maxmimum number of rows! <br>",  $file_specifier);
                        break;
                    }
                    if ($total_income != '')
                    {
                        printer("<br>All data gathered successfully, exiting!<br>",  $file_specifier);
                        break;
                    }

                    
                    $summary = '';
                    $task_name = '';
                    $material_type = '';
                    $material_amount = '';
                    $material_units = '';
                    $material_unit_price = '';
                    $work_duration = '';
                    $work_units = ''; 
                    $work_amount = '';

                    //Find header
                    if ($c[0] == 'Eil. Nr.')
                    {
                        $queue_numb = $crr_row_idx;
                        printer("Found header! <br>",  $file_specifier);
                        $y = 0;
                        for ($y; $y <= $num_cols; $y++)
                        {  
                            // Find measurement column in the header
                            if (isset($c[$y]) && $c[$y] == 'Mato vnt.')
                            {
                                $measurement_col_numb = $y;
                                continue;
                            }else {}

                            // Find price column in the header
                            if (isset($c[$y]) && $c[$y] == 'Kaina')
                            {
                                $price_col_numb = $y;
                                continue;
                            }else{}

                            // Find work column in the header
                            if (isset($c[$y]) && $c[$y] == 'Darbas')
                            {
                                $work_col_numb = $y;
                                continue;
                            }else{}

                            // Find material_amoount column in the header
                            if (isset($c[$y]) && $c[$y] == 'Medžiagos')
                            {
                                $material_amoount_col_numb = $y;
                                continue;
                            }else{}

                            // Find amount column in the header
                            if (isset($c[$y]) && $c[$y] == 'Kiekis')
                            {
                                $amount = $y;
                                continue;
                            }else {}

                            // Find mechanism column in the header
                            if (isset($c[$y]) && $c[$y] == 'Mechanizmai')
                            {
                                $mecha1 = $y;
                                continue;
                            }else {}

                            // Find mechanism column in the header
                            if (isset($c[$y]) && $c[$y] == 'Irengimai')
                            {
                                $mecha2 = $y;
                                continue;
                            }else {}

                            // Find sum column in the header
                            if (isset($c[$y]) && $c[$y] == 'Suma')
                            {
                                $sum = $y;
                                continue;
                            }else {}

                        }
                    }

                    // Finds estimate sum info variables
                    if ($c[1] == 'Statinio statybos išlaidos'){
                        $work_total_sum = $c[$work_col_numb];
                        $material_total_sum = $c[$material_amoount_col_numb];
                        $tools_total_sum = $c[$mecha1];
                        $tools2_total_sum = $c[$mecha2];
                    }
                    
                    if ($c[2] == 'Statybvietės išlaidos'){$expences1 = $c[$sum];}
                    if ($c[2] == 'Pridėtinės išlaidos'){$expences2 = $c[$sum];}
                    if ($c[2] == 'Kita'){$expences3 = $c[$sum];}
                    if ($c[2] == 'Pelnas'){$profit = $c[$sum];}
                    if ($c[2] == 'Pridėtinės vertės mokestis'){$total_income = $c[$sum];}


                    // Check if variables have been assigned and put them into an array
                    if ($work_total_sum != '' && $total_income != ''){
                        $estimate_sum_info = ["work_total_sum"=>$work_total_sum, "material_total_sum"=>$material_total_sum,
                                                "tools_total_sum"=>$tools_total_sum, "tools2_total_sum"=>$tools2_total_sum,
                                                "expences1"=>$expences1, "expences2"=>$expences2, "expences3"=>$expences3,
                                                "profit"=>$profit, "total_income"=>$total_income];
                        array_push($estimate_sum_array, $estimate_sum_info);
                    }

                    // Finds the task group name variable
                    if ($work_col_numb != $invalid_val){
                        if (gettype($c[0]) == 'string'){
                            if ($c[0] == '' && $c[$work_col_numb] != '' && $c[$material_amoount_col_numb] != '' && $c[$mecha1] != ''){
                                $task_group_name = $c[1];
                            }
                        }
                    }

                    //Check if summaries first cell has a digit and if its a floating number
                    if ($crr_row_idx > $queue_numb && $c[0] == (float)$c[0] && is_numeric($c[0]))
                    {   
                        //Check if header was found
                        if ($work_col_numb == $invalid_val)
                        {
                            printer('Header not found<br>',  $file_specifier);
                            continue;
                        }
                        $jobs_found++;
                        printer("<br>Found job $jobs_found! <br>",  $file_specifier);
                        
                        //Assign summary number
                        $summary = $c[0];

                        //Assign name variable
                        $task_name = $c[1];
                        $work_units = $c[$measurement_col_numb];
                        $work_amount = $c[$amount];
                        if ($file_specifier == 1)
                        {
                            print_r($task_name);
                        }
                        printer('<br>', $file_specifier);
                        $next_row_idx = $crr_row_idx+1;
                        $next_row = $rows[$next_row_idx];

                        //Find and assign work variables
                        if ($next_row[0] == '' && $next_row[$work_col_numb] != '')
                        {
                            $work_duration = $next_row[$amount];
                            //$work_price = $next_row[$price_col_numb];

                        }else
                        {
                            printer("No work data <br>",  $file_specifier);
                        }
                        
                        $next_row_idx++;
                        $next_row = $rows[$next_row_idx];

                        //Find and assign material variables
                        if ($next_row[0] == '' && $next_row[$material_amoount_col_numb] != '')
                        {
                            $material_type = $next_row[1];
                            $material_amount = $next_row[$amount];
                            $material_units = $next_row[$measurement_col_numb];
                            $material_unit_price = $next_row[$price_col_numb];
                            
                        }else{printer("No material data <br>",  $file_specifier);}

                        //Insert into array
                        if ($work_duration != '')
                        {
                            $new_jobs = ["task_group_name"=>$task_group_name, "line_num"=>$summary, "job_name"=>$task_name, "material_name"=>$material_type,
                                     "material_amount"=>$material_amount, "material_amoount_unit"=>$material_units, "material_amoount_unit_price"=>$material_unit_price,
                                     "work_amount"=>$work_amount,"work_unit"=>$work_units, "work_duration"=>$work_duration];
                            array_push($jobs_array, $new_jobs);
                        }
                        
                            
                    }else{}
                    
                    
                    $crr_row_idx++;
                }

                printer( '<br>Finished!<br>',  $file_specifier);
                printer( 'Found '.$jobs_found.' jobs!<br>',  $file_specifier);

                // Makes a new array, combining the previous 2 arrays
                

                // Prints out the array
                if ($file_specifier == 1)
                {
                    $group_array = [];
                    array_push($group_array, $jobs_array);
                    array_push($group_array, $estimate_sum_array);

                    echo '<pre>';
                    print_r($group_array);
                    echo '</pre>';

                    return $group_array;
                }else if ($file_specifier == 0){
                    return $jobs_array;
                }
            }
        }
    }
?>